File: gradle/subprojects/numbers-utils/src/main/kotlin/com/prog/numbers/utils/app/Factor.kt
package com.prog.numbers.utils.app

import com.prog.numbers.utils.operations.FactorUtils.factor
import com.prog.numbers.utils.operations.FactorUtils.primeFactor
import java.math.BigInteger

/**
 * # `getUserNumber`
 *
 * This function prompts the user to enter a number.
 *
 * __Return:__
 *
 * * `BigInteger` - The number entered by the user or from the command line
 * arguments
 *
 * __Example:__
 *
 * ```kotlin
 * import com.prog.numbers.utils.app.getNumber
 * val number = getNumber()
 * println("Number : $number")
 * ```
 *
 * __Output:__
 *
 * ```kotlin
 * "Enter a number : 5"
 * "Number : 5"
 * ```
 */
private fun getUserNumber(): BigInteger {
    print("Enter a number : ")
    val numberStr = readLine() ?: ""
    return numberStr.toBigIntegerOrNull() ?: BigInteger.ZERO
}

/**
 * # `getNumber`
 *
 * This function returns the number provided by the user or from the command line
 *
 * __Arguments:__
 *
 * * `args: Array<String>` - An array of strings that may contain the user's
 * number
 *
 * __Return:__
 *
 * * `BigInteger` - The number provided by the user or from the command line
 * arguments
 */
fun getNumber(args: Array<String>): BigInteger {
    return if (args.isNotEmpty() && args[0].toBigIntegerOrNull() != null) {
        args[0].toBigIntegerOrNull()!!
    } else {
        getUserNumber()
    }
}

/**
 * # `printFactor`
 *
 * This function prints the factors of a given number.
 *
 * __Arguments:__
 *
 * * `number: BigInteger` - The number for which the factors need to be printed
 */
fun printFactor(number: BigInteger) {
    val factors = factor(number)
    val factorsStr = factors.toString()
    println("Factor[$number] = $factorsStr")
}

/**
 * # `printPrimeFactor`
 *
 * This function prints the prime factors of a given number.
 *
 * __Arguments:__
 *
 * * `number: BigInteger` - The number for which the prime factors need to be
 * printed
 */
fun printPrimeFactor(number: BigInteger) {
    val primeFactors = primeFactor(number)
    val primeFactorsStr = primeFactors.toString()
    println("PFactor[$number] = $primeFactorsStr")
}

/**
 * # `main`
 *
 * This function is the entry point of the application.
 *
 * __Arguments:__
 *
 * * `args: Array<String>` - Command line arguments
 */
fun main(args: Array<String>) {
    val number = getNumber(args)
    printFactor(number)
    printPrimeFactor(number)
}

File: haskell/packages/numbers-utils/app/Factor.hs
module Main where

import Data.List (intercalate)
import Data.Maybe (fromJust, fromMaybe, isJust)
import Numeric.Natural (Natural)
import System.Environment (getArgs)
import System.IO (hFlush, stdout)
import Text.Read (readMaybe)

import Numbers.Utils.Operations.FactorUtils (factor, primeFactor)

-- |
-- = @getUserNumber@
--
-- The function @getUserNumber@ is used to get a number from the user.
--
-- __Return:__
--
-- * @Natural@ - The number entered by the user.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FactorUtils (getUserNumber)
-- number <- getUserNumber
-- putStrLn $ "Number : " ++ show number
-- @
--
-- __Output:__
--
-- @
-- "Enter a number : 5"
-- "Number : 5
-- @
getUserNumber :: IO Natural
getUserNumber = do
    putStr "Enter a number : "
    hFlush stdout
    fromMaybe 0 . readMaybe <$> getLine

-- |
-- = @getNumber@
--
-- The function @getNumber@ is used to get a number from the command line
-- arguments.
--
-- __Arguments:__
--
-- * @args: IO@ - The command line arguments.
--
-- __Return:__
--
-- * @Natural@ - The number entered by the user.
getNumber :: IO Natural
getNumber = do
    args <- getArgs
    if not (null args) && isJust (readMaybe (head args) :: Maybe Natural)
        then return $ fromJust (readMaybe (head args) :: Maybe Natural)
        else getUserNumber

-- |
-- = @printFactor@
--
-- The function @printFactor@ is used to print the factors of a number.
--
-- __Arguments:__
--
-- * @Natural@ - The number.
--
-- __Return:__
--
-- * @IO ()@ - The factors of the number.
printFactor :: Natural -> IO ()
printFactor number = do
    let factors = factor number
    let factorsStr = intercalate ", " (map show factors)
    putStrLn $
        "Factor["
            ++ show number
            ++ "] = ["
            ++ factorsStr
            ++ "]"

-- |
-- = @printPrimeFactor@
--
-- The function @printPrimeFactor@ is used to print the prime factors of a
-- number.
--
-- __Arguments:__
--
-- * @Natural@ - The number.
--
-- __Return:__
--
-- * @IO ()@ - The prime factors of the number.
printPrimeFactor :: Natural -> IO ()
printPrimeFactor number = do
    let primeFactors = primeFactor number
    let primeFactorsStr = intercalate ", " (map show primeFactors)
    putStrLn $
        "PFactor["
            ++ show number
            ++ "] = ["
            ++ primeFactorsStr
            ++ "]"
-- |
-- = @main@
--
-- The function @main@ is the entry point of the program. It gets a number from
-- the user and prints the factors and prime factors of the number.
main :: IO ()
main = do
    number <- getNumber
    --  Find faster way to print factors
    printFactor number
    printPrimeFactor number
